### PR TITLE
Mini-Challenge 1: Core Concepts and Styling

### PR DESCRIPTION
**Pull request by Fetrelo**
Removed base code
Added basic header UI
Added video grid
Added basic responsiveness
Added initial styling to overall app

### CODE REVIEW

**Comment by ragrum** [comment]
Please, remember that mini-challenges must be worked on a branch based off your own `master/main` branch in the forked repo under your own GitHub account.

Then, deliverable pull-requests must be open against your own forked repo's `master/main` branch, so that pull-requests may be merged and integrated for future mini-challenges to build on top of that.

For this time, since you've got all your commits already in your forked `master/main` branch I'll close this PR without merging after all the feedback is collected and then you should continue working as stated above.

File: src/components/CustomInputs/InputTextIcon.jsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      <InputWrapper>
+        <StyledInputText />
+      </InputWrapper>
+    </Wrapper>
+  );
+};
+
+export default InputTextIcon;
+
+const Wrapper = styled.div`
```
Comment (ragrum) [inline src/components/CustomInputs/InputTextIcon.jsx:23]: A common practice is to define all your `styled` component above the default exported component (the one with the `render` method).  No change is required as this  as this is a preference.

File: src/components/CustomInputs/InputTextIcon.jsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
+import React from 'react';
+import styled from 'styled-components';
+
+const InputTextIcon = ({ path }) => {
+  return (
+    <Wrapper data-testid="searchTextWrapper">
+      <SvgWrapper>
+        <StyledSvg focusable="false" viewBox="0 0 24 24">
+          {path || (
+            <path d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z" />
```
Comment (ragrum) [inline src/components/CustomInputs/InputTextIcon.jsx:10]: You could use [SVGr](https://react-svgr.com/) to convert SVGs to React components. No change is required as this as this is a preference.

**Comment by ragrum** [comment]
Just being curious, what does this button will do?

image

Is the one to log in and log out a user?

Edit: NVM, I just found `<StyledLoginSvg />` 🙂

File: src/components/CustomInputs/Toggle.jsx [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import React, { useState } from 'react';
+import styled from 'styled-components';
+
+const Toggle = ({ labelOn, labelOff }) => {
+  const [isChecked, setIsChecked] = useState(false);
+
+  const handleCheck = (e) => {
```
Comment (ragrum) [inline src/components/CustomInputs/Toggle.jsx:7]: You can use the component state itself for the toggling:

```js
const handleCheck = () => {
  setIsChecked(!isChecked);
};
```

File: src/components/Nav.jsx [hunk 4]
```diff
@@ -0,0 +1,90 @@
+import React from 'react';
+import styled from 'styled-components';
+import InputTextIcon from './CustomInputs/InputTextIcon';
+import Toggle from './CustomInputs/Toggle';
+
+const Nav = () => {
+  return (
+    <StyledNav data-testid="homepage">
```
Comment (ragrum) [inline src/components/Nav.jsx:8]: Remember that the `data-testid` attribute must be used sparingly. Depending on your testing strategy you may want to remove these for the production build.

File: src/components/VideoCard.jsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
+import React from 'react';
+import styled from 'styled-components';
+
+const VideoCard = ({ data }) => {
```
Comment (ragrum) [inline src/components/VideoCard.jsx:4]: A common practice is to pass a set of properties instead of whole objects, you can destructure them the moment they're read `{ data: etag, snippet: { publishedAt, thumbnails, ...  } }`

File: src/pages/Home/Home.page.jsx [hunk 6]
```diff
@@ -1,39 +1,44 @@
...
-        <Link to="/login">let me in →</Link>
-      )}
-    </section>
+    <>
+      <Nav />
+      <BodyContainer>
+        <h1>YouTube video search app</h1>
+        <VideoGrid>
+          {youtubeVideos.items
+            .filter((obj) => obj.id.videoId && true)
```
Comment (ragrum) [inline src/pages/Home/Home.page.jsx:15]: Perhaps a better name for each item in `youtubeVideos` would be `video` instead of `obj`?:

```
.filter((video) => video.id.videoId && true)
```

File: src/__tests__/homepage.test.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
+import React from 'react';
+import '@testing-library/jest-dom';
+import {cleanup, render, screen} from '@testing-library/react';
+import HomePage from "../pages/Home/Home.page";
+import youtubeVideos from "../mocks/youtube-videos";
+
+describe('Home page Tests', () => {
+  afterEach(cleanup);
```
Comment (ragrum) [inline src/__tests__/homepage.test.js:8]: This is [done automatically](https://testing-library.com/docs/react-testing-library/setup/#skipping-auto-cleanup) for you when using `Jest`.

File: src/__tests__/homepage.test.js [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+  describe('Ensure elements get rendered', () => {
+    it('Title', () => {
+      render(<HomePage />);
+      expect(screen.getByText(/YouTube video search app/, {selector: 'h1'})).toBeInTheDocument();
```
Comment (ragrum) [inline src/__tests__/homepage.test.js:13]: Prefer the `getByRole` query in this case with `heading` role.

File: src/__tests__/homepage.test.js [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    });
+  });
+
+  describe('Ensure there the info from the mock is being shown', () => {
+    it("A random video's information will be displayed if it has videoid", () => {
```
Comment (ragrum) [inline src/__tests__/homepage.test.js:18]: Perhaps implementing the video cards using the `article` HTML element may allow you to use the `getByRole` with role equals to `article` and `name` equals to a random video title?

File: src/__tests__/homepage.test.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+      const randomVideo = youtubeVideos.items[randVideoIdx];
+
+      if(randomVideo.id.videoId)
+        expect(screen.getByTestId(randomVideo.etag)).toBeInTheDocument();
+      else
+        expect(screen.getByTestId(randomVideo.etag)).not.toBeInTheDocument();
+    });
+
+    it('A video that does not have any description, must show channel title followed by a hypen', () => {
+      // render(<HomePage />);
```
Comment (ragrum) [inline src/__tests__/homepage.test.js:30]: Unimplemented tests are better off from the commit.

File: src/__tests__/nav.test.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
+import React from 'react';
+import '@testing-library/jest-dom';
+import {cleanup, fireEvent, render, screen} from '@testing-library/react';
+import Toggle from '../components/CustomInputs/Toggle';
+import Nav from '../components/Nav';
+
+describe('Nav tests', () => {
+  afterEach(cleanup);
+
+  describe('Ensure components get rendered', () => {
```
Comment (ragrum) [inline src/__tests__/nav.test.js:10]: Instead of having lots of shorts tests you may try [with longer](https://kentcdodds.com/blog/write-fewer-longer-tests) tests.

**Comment by ragrum** [comment]
> Run test coverage and save results into a .txt file at ./ - root level. Can you make the total coverage to be above of 60%?

The coverage file is missing.

```bash
npx jest --coverage > coverage.txt
```

**Comment by ragrum** [comment]
While running the tests `npx jest` I once got a failing tests, it seems it's due to the random video id use, couldn't reproduce it again.

File: src/components/CustomInputs/InputTextIcon.jsx (hunk 1 above)
Comment (Fetrelo) [reply src/components/CustomInputs/InputTextIcon.jsx:23]: Thank you! I will change it in the next deliverable

File: src/components/CustomInputs/InputTextIcon.jsx (hunk 2 above)
Comment (Fetrelo) [reply src/components/CustomInputs/InputTextIcon.jsx:10]: Oh ok, I didn't know about that tool. Thanks!

File: src/components/CustomInputs/Toggle.jsx (hunk 3 above)
Comment (Fetrelo) [reply src/components/CustomInputs/Toggle.jsx:7]: I thought that it would be sort of "more reliable" if I got the actual value from the input directly. As if, for some reason, the value of the checkbox gets updated without having called the onChange event just once, or something...

File: src/components/Nav.jsx (hunk 4 above)
Comment (Fetrelo) [reply src/components/Nav.jsx:8]: Ok, got it! Thanks :)

File: src/pages/Home/Home.page.jsx (hunk 6 above)
Comment (Fetrelo) [reply src/pages/Home/Home.page.jsx:15]: Absolutely! Haha I thought I had changed it

File: src/__tests__/homepage.test.js (hunk 7 above)
Comment (Fetrelo) [reply src/__tests__/homepage.test.js:8]: Hahaha I had read that but the anxiety prevented me from avoiding it. Thanks

File: src/__tests__/homepage.test.js (hunk 8 above)
Comment (Fetrelo) [reply src/__tests__/homepage.test.js:13]: Thanks, I will change it.

**Comment by Fetrelo** [comment]
> > Run test coverage and save results into a .txt file at ./ - root level. Can you make the total coverage to be above of 60%?
> 
> The coverage file is missing.
> 
> ```shell
> npx jest --coverage > coverage.txt
> ```

[coverage.txt](https://github.com/erickwize/react-certification-2020/files/6207501/coverage.txt)

image

Here they are

**Comment by Fetrelo** [comment]
> While running the tests `npx jest` I once got a failing tests, it seems it's due to the random video id use, couldn't reproduce it again.

Oh, perhaps. What should I do? Refactor that test?

File: src/__tests__/homepage.test.js (hunk 10 above)
Comment (Fetrelo) [reply src/__tests__/homepage.test.js:30]: We can review that test to see how could it be best approached.

File: src/components/VideoCard.jsx (hunk 5 above)
Comment (Fetrelo) [reply src/components/VideoCard.jsx:4]: Got it, thanks!

**Comment by ragrum** [comment]
> > While running the tests `npx jest` I once got a failing tests, it seems it's due to the random video id use, couldn't reproduce it again.
> 
> Oh, perhaps. What should I do? Refactor that test?

That would be a good idea, if this were a production project that test would be flaky and might fail from time to time, making your automated tests unreliaable.

File: src/components/CustomInputs/Toggle.jsx (hunk 3 above)
Comment (ragrum) [reply src/components/CustomInputs/Toggle.jsx:7]: Check these articles:

https://reactjs.org/docs/forms.html#controlled-components
https://reactjs.org/docs/uncontrolled-components.html

TL;DR: Common practice is to make your form inputs be controlled with React state.
